### PR TITLE
Fix indexing integer timestamp, refactor with `timestamp_range`

### PR DESCRIPTION
**Pull request by d-a-bunin**
## Before submitting (must do checklist)

- [x] Did you read the [contribution guide](https://github.com/etna-team/etna/blob/master/CONTRIBUTING.md)?
- [x] Did you update the docs? We use Numpy format for all the methods and classes.
- [x] Did you write any new necessary tests?
- [x] Did you update the [CHANGELOG](https://github.com/etna-team/etna/blob/master/CHANGELOG.md)?

## Proposed Changes
- Add `timestamp_range` and refactor code
- Add fixes for some bugs

### CODE REVIEW

File: etna/datasets/utils.py [hunk 1]
```diff
@@ -426,3 +426,61 @@ def determine_freq(timestamps: Union[pd.Series, pd.Index]) -> Optional[str]:
...
+        raise ValueError("Of the three parameters: start, end, periods, exactly two must be specified!")
+
+    if freq is None:
+        if start is None:
+            start = end - periods + 1  # type: ignore
+        if periods is None:
+            periods = end - start + 1  # type: ignore
+        return pd.Index(np.arange(start, start + periods))
+    else:
+        return pd.date_range(start=start, end=end, periods=periods, freq=freq)
```
Comment (alex-hse-repository) [inline etna/datasets/utils.py:486]: In some places we used `closed="right"` parameter, it it equivalent to not set it?

File: etna/transforms/math/differencing.py [hunk 2]
```diff
@@ -194,12 +199,9 @@ def _reconstruct_test(self, df: pd.DataFrame, columns_to_inverse: Set[str]) -> p
         result_df = df.copy()
 
         # check that test is right after the train
-        expected_min_test_timestamp = pd.date_range(
-            start=self._test_init_df.index.max(),  # type: ignore
-            periods=2,
-            freq=pd.infer_freq(self._train_timestamp),
-            closed="right",
-        )[0]
+        expected_min_test_timestamp = timestamp_range(
```
Comment (alex-hse-repository) [inline etna/transforms/math/differencing.py:202]: Doesn't look the same as previous version

File: etna/transforms/math/differencing.py [hunk 3]
```diff
@@ -398,6 +402,9 @@ def _fit(self, df: pd.DataFrame) -> "DifferencingTransform":
         result_df = df
         for transform in self._differencing_transforms:
             result_df = transform._fit_transform(result_df)
+            transform._freq = self._freq
```
Comment (alex-hse-repository) [inline etna/transforms/math/differencing.py:405]: Do we need it here?

File: etna/transforms/missing_values/resample.py [hunk 4]
```diff
@@ -61,8 +61,8 @@ def _get_folds(self, df: pd.DataFrame) -> List[int]:
         n_periods = len(df) // n_folds_per_gap + 2
 
         in_column_start_index = in_column_index[0]
-        left_tie_len = len(df[:in_column_start_index]) - 1
-        right_tie_len = len(df[in_column_start_index:])
+        left_tie_len = len(df.loc[:in_column_start_index]) - 1
```
Comment (alex-hse-repository) [inline etna/transforms/missing_values/resample.py:64]: Do we have tests for this?)

File: tests/test_transforms/test_inference/conftest.py [hunk 5]
```diff
@@ -169,21 +169,21 @@ def ts_to_resample() -> TSDataset:
 def ts_to_resample_int_timestamp() -> TSDataset:
     df_1 = pd.DataFrame(
         {
-            "timestamp": np.arange(120),
+            "timestamp": np.arange(24, 144),
```
Comment (alex-hse-repository) [inline tests/test_transforms/test_inference/conftest.py:172]: Why did you change it?

File: etna/analysis/eda/plots.py [hunk 6]
```diff
@@ -341,7 +341,7 @@ def plot_holidays(
 
     for i, segment in enumerate(segments):
         segment_df = df.loc[start:end, pd.IndexSlice[segment, "target"]]  # type: ignore
-        segment_df = segment_df[segment_df.first_valid_index() : segment_df.last_valid_index()]
+        segment_df = segment_df.loc[segment_df.first_valid_index() : segment_df.last_valid_index()]
```
Comment (alex-hse-repository) [inline etna/analysis/eda/plots.py:344]: How can we find all such places? If it possible to handle the search via type checking or linters? Looks like we miss many of them

File: etna/analysis/forecast/plots.py [hunk 7]
```diff
@@ -430,7 +431,7 @@ def plot_backtest_interactive(
         for fold_number in folds:
             start_fold = fold_numbers[fold_numbers == fold_number].index.min()
             end_fold = fold_numbers[fold_numbers == fold_number].index.max()
-            end_fold_exclusive = pd.date_range(start=end_fold, periods=2, freq=ts.freq)[1]
+            end_fold_exclusive = timestamp_range(start=end_fold, periods=2, freq=ts.freq)[-1]
```
Comment (alex-hse-repository) [inline etna/analysis/forecast/plots.py:434]: Why is it -1 now?

**Review by alex-hse-repository — CHANGES_REQUESTED** [review]

File: etna/analysis/eda/plots.py (hunk 6 above)
Comment (d-a-bunin) [reply etna/analysis/eda/plots.py:344]: I don't know how to find them by using type checking of linters. I searched for `df[` in a repository.

File: etna/analysis/forecast/plots.py (hunk 7 above)
Comment (d-a-bunin) [reply etna/analysis/forecast/plots.py:434]: Because there are only 2 elements there with indices 0 and 1.  It has no difference.

File: etna/datasets/datasets_generation.py [hunk 8]
```diff
@@ -9,20 +9,18 @@
...
-        return np.arange(start_time, start_time + periods)  # type: ignore
-    else:
-        if start_time is None:
-            start_time = pd.Timestamp("2000-01-01")
-        return pd.date_range(start=start_time, freq=freq, periods=periods)
+    if freq is None and start_time is None:
+        start_time = 0
+    if freq is not None and start_time is None:
+        start_time = pd.Timestamp("2000-01-01")
+    _check_timestamp_param(param=start_time, param_name="start_time", freq=freq)
```
[the remark this answers is not in the document]
Comment (d-a-bunin) [reply etna/datasets/datasets_generation.py:22]: In `timestamp_range` the error message will be about parameter `start` and here we have `start_time`.

File: etna/datasets/tsdataset.py [hunk 9]
```diff
@@ -260,18 +261,11 @@ def __getitem__(self, item):
...
-        if freq is None:
-            new_index = np.arange(df.index.min(), df.index.max() + future_steps + 1)
-        else:
-            with warnings.catch_warnings():
-                warnings.filterwarnings(action="ignore", message="Argument `closed` is deprecated")
-                future_dates = pd.date_range(start=df.index.max(), periods=future_steps + 1, freq=freq, closed="right")
-            new_index = df.index.append(future_dates)
-
+        to_add_index = timestamp_range(start=df.index[-1], periods=future_steps + 1, freq=freq)[1:]
+        new_index = df.index.append(to_add_index)
```
[the remark this answers is not in the document]
Comment (d-a-bunin) [reply etna/datasets/tsdataset.py:265]: In previous version for integer timestamp we haven't used `append` to current index.

File: etna/datasets/utils.py (hunk 1 above)
Comment (d-a-bunin) [reply etna/datasets/utils.py:486]: This version generates all timestamps from `start` to `end` including both ends. If we want to exclude one of the ends we could just use `[:-1]` or `[1:]` on the result of `timestamp_range`.

File: etna/transforms/missing_values/resample.py (hunk 4 above)
Comment (d-a-bunin) [reply etna/transforms/missing_values/resample.py:64]: I changed fixture for resampling in inference tests to check this.

File: tests/test_datasets/test_utils.py [hunk 10]
```diff
@@ -458,3 +459,58 @@ def test_determine_freq_fail_cant_determine(timestamps):
...
+)
+def test_timestamp_range(start, end, periods, freq, expected_range):
+    result = timestamp_range(start=start, end=end, periods=periods, freq=freq)
+    np.testing.assert_array_equal(result, expected_range)
+
+
+@pytest.mark.parametrize(
+    "start, end, periods, freq",
+    [
+        ("2020-01-01", "2020-01-10", None, None),
```
[the remark this answers is not in the document]
Comment (d-a-bunin) [reply tests/test_datasets/test_utils.py:484]: We understand that type is incorrect based on given `freq`. If `freq` is None type should be int, if `freq` is str type should be str/timestamp.

File: tests/test_transforms/test_inference/conftest.py (hunk 5 above)
Comment (d-a-bunin) [reply tests/test_transforms/test_inference/conftest.py:172]: We changed it to check that resampling uses `.loc`.

File: etna/transforms/math/differencing.py (hunk 2 above)
Comment (d-a-bunin) [reply etna/transforms/math/differencing.py:202]: The difference:
- Add using `self._freq` saved during `fit`
- Remove `closed="right"` and use `[-1]`
- Use `self._test_init_df.index[-1]` instead of `self._test_init_df.index.max()`

File: etna/transforms/math/differencing.py (hunk 3 above)
Comment (d-a-bunin) [reply etna/transforms/math/differencing.py:405]: Here we fit `_SingleDifferencingTransform` on `pd.DataFrame`, but it can't get `_freq` from `pd.DataFrame` and we should somehow pass it into it.

File: etna/datasets/tsdataset.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (alex-hse-repository) [reply etna/datasets/tsdataset.py:265]: So was it bug?)

File: etna/datasets/utils.py (hunk 1 above)
Comment (alex-hse-repository) [reply etna/datasets/utils.py:486]: So it is not equivalent?)

File: etna/transforms/math/differencing.py (hunk 2 above)
Comment (alex-hse-repository) [reply etna/transforms/math/differencing.py:202]: Don't get why we use [-1] instead of [0]

File: etna/transforms/math/differencing.py (hunk 3 above)
Comment (alex-hse-repository) [reply etna/transforms/math/differencing.py:405]: Here we also define this parameter
Снимок экрана 2024-02-22 в 13 23 10

File: etna/analysis/forecast/plots.py (hunk 7 above)
Comment (alex-hse-repository) [reply etna/analysis/forecast/plots.py:434]: Then it is better to leave the old logic

**Review by alex-hse-repository — APPROVED** [review]
